File: Learn_Basic/src/learn_py/learn_py/opencv_find_point.py
import cv2
import numpy as np
import rclpy
from rclpy.node import Node

def Find_Point(origin):
    # 将图像从BGR转换为HSV颜色空间
    hsv = cv2.cvtColor(origin, cv2.COLOR_BGR2HSV)

    # 定义红色的HSV范围
    hsv_lower = np.array([156, 43, 46])
    hsv_upper = np.array([180, 255, 255])
    mask = cv2.inRange(hsv, hsv_lower, hsv_upper)
    kernel = np.ones((2, 2), np.uint8)
    mask_opened = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)
    contours, _ = cv2.findContours(mask_opened, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)

    # 处理查找结果
    radius_max = 0
    point_result = []
    for cnt in contours:
        epsilon = 0.1 * cv2.arcLength(cnt, True)
        approx = cv2.approxPolyDP(cnt, epsilon, True)
        # 拟合
        (x, y), radius = cv2.minEnclosingCircle(approx)
        if radius_max < radius:
            radius_max = radius
            point_result.append((int(x), int(y), int(radius)))  # 添加半径到结果中
            
    return radius_max != 0, point_result


def draw_point_circle(image, posArray):
    for pos in posArray:
        center = (pos[0], pos[1])
        radius = pos[2]  # 使用传入的半径
        cv2.circle(image, center, radius, (255, 0, 0), 2)
        
def main(args = None):
    
    rclpy.init(args=args)
    
    node = Node("opencv_node")
    
    cap = cv2.VideoCapture(0)

    while True:
        ret, frame = cap.read()
        if not ret:
            break
        try:
            isFind, points = Find_Point(frame)
            if isFind:
                draw_point_circle(frame, points)
                node.get_logger().info(f"Found {len(points)} points : {points}")

            cv2.imshow("photo", frame)
        except Exception as e:
            print(e)  # 打印异常信息
            continue

        # 按下'q'键退出循环
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    # 释放摄像头资源
    cap.release()
    cv2.destroyAllWindows()
    rclpy.shutdown()

File: Learn_Basic/src/learn_cmake/src/opencv_find_point.cpp
#include "rclcpp/rclcpp.hpp"
#include <opencv2/opencv.hpp>
#include <vector>
#include <chrono>
#include <thread>

std::pair<bool, std::vector<std::tuple<int, int, float>>> Find_Point(const cv::Mat &origin)
{
    cv::Mat hsv;
    cv::cvtColor(origin, hsv, cv::COLOR_BGR2HSV);

    // Define HSV range for red
    cv::Scalar hsv_lower(60, 43, 46);
    cv::Scalar hsv_upper(77, 255, 255);

    cv::Mat mask;
    cv::inRange(hsv, hsv_lower, hsv_upper, mask);

    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2, 2));
    cv::morphologyEx(mask, mask, cv::MORPH_OPEN, kernel);

    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_NONE);

    int radius_max = 0;
    std::vector<std::tuple<int, int, float>> point_result;

    for (const auto &cnt : contours)
    {
        double epsilon = 0.1 * cv::arcLength(cnt, true);
        std::vector<cv::Point> approx;
        cv::approxPolyDP(cnt, approx, epsilon, true);

        cv::Point2f center;
        float radius;
        cv::minEnclosingCircle(approx, center, radius);

        if (radius_max < radius)
        {
            radius_max = radius;
            point_result.emplace_back(
                static_cast<int>(center.x),
                static_cast<int>(center.y),
                radius);
        }
    }

    return {radius_max != 0, point_result};
}

void draw_point_circle(cv::Mat &image, const std::vector<std::tuple<int, int, float>> &posArray)
{
    for (const auto &pos : posArray)
    {
        cv::Point center(std::get<0>(pos), std::get<1>(pos));
        int radius = std::get<2>(pos);
        cv::circle(image, center, radius, cv::Scalar(255, 0, 0), 2);
    }
}

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);

    cv::VideoCapture cap(0);
    if (!cap.isOpened())
    {
        std::cerr << "Error opening video capture" << std::endl;
        return EXIT_FAILURE;
    }

    cv::Mat frame;
    auto node = std::make_shared<rclcpp::Node>("hello_world_test_node");

    while (rclcpp::ok())
    {
        cap >> frame;
        if (frame.empty())
            break;

        try
        {
            auto [isFind, points] = Find_Point(frame);

            if (isFind)
            {
                std::string outputInfo;

                for (const auto &point : points)
                {
                    outputInfo += "Found " + std::to_string(points.size()) + " points : ";
                    outputInfo += "(" + std::to_string(std::get<0>(point)) + ", " + std::to_string(std::get<1>(point)) + ", " + std::to_string(std::get<2>(point)) + ") ";
                }

                RCLCPP_INFO(node->get_logger(), "%s", outputInfo.c_str());
            }

            if (isFind)
            {
                draw_point_circle(frame, points);
            }

            cv::imshow("photo", frame);
        }
        catch (const cv::Exception &e)
        {
            std::cerr << "OpenCV exception: " << e.what() << std::endl;
            continue;
        }

        if (cv::waitKey(1) == 'q')
        {
            break;
        }
    }
    cap.release();
    cv::destroyAllWindows();

    rclcpp::shutdown();
    return 0;
}
